blob/azureblob/azureblob.go

```diff
@@ -816,14 +828,48 @@ func (b *bucket) ListPaged(ctx context.Context, opts *driver.ListOptions) (*driv
 	return page, nil
 }
 
+func (b *bucket) getDelegationCredentials(ctx context.Context) (*azblob.UserDelegationCredential, error) {
+	validPeriod := 48 * time.Hour
+	currentTime := time.Now().UTC()
+	expires := currentTime.Add(validPeriod)
+	b.credentialExpiration = expires
+	keyInfo := azblob.NewKeyInfo(currentTime, expires)
+	delegationCredentials, err := b.serviceURL.GetUserDelegationCredential(ctx, keyInfo, nil, nil)
+
+	if err != nil {
+		return nil, fmt.Errorf("error while retrieving user delegation credential: %s", err)
+	}
+
+	return &delegationCredentials, nil
+}
+
 // SignedURL implements driver.SignedURL.
 func (b *bucket) SignedURL(ctx context.Context, key string, opts *driver.SignedURLOptions) (string, error) {
+	var credential azblob.StorageAccountCredential
 	if b.opts.Credential == nil {
-		return "", gcerr.New(gcerr.Unimplemented, nil, 1, "azureblob: to use SignedURL, you must call OpenBucket with a non-nil Options.Credential")
+		var err error
```

Comment:
I think this could be simplified to something like:

```
var credential azblob.StorageAccountCredential
if b.opts.Credential != nil {
  credential = b.opts.Credential
} else if isMSIEnvironment {
  var err error
  credential, err = b.refreshDelegationCredentials(ctx)
  if err != nil { ...}
} else {
  // return error
}
// ... use credential
```

where `refreshDelegationCredentials` checks the expiry and refreshes if needed.

`SignedURL` can be called by multiple goroutines, so b.delegationCredentials and b.credentialExpiration need to be protected by a mutex; doing all reading/writing of those variables inside a single function makes that easy.